### PR TITLE
Decode the user info of a database URI

### LINKED ISSUE #3716: The user info part of an URI should be decoded
**Issue type:**
Bug report

**Database system/driver:**
- `mysql` / `mariadb`
- `postgres`

**TypeORM version:**
`0.2.14`

**Steps to reproduce or a small repository showing the problem:**
If I use this `ormconfig.yaml` file, I can connect to my database:

```yaml
default:
  type: postgres
  host: localhost
  port: 5432
  database: foo_bar
  username: postgres
  password: Foo@Bar@      # Note the "at" signs!
```

But if I use an `url` instead, my password is refused:

```yaml
default:
  type: postgres
  url: 'postgres://postgres:Foo%40Bar%40@localhost:5432/foo_bar'
```

The password sent to the database is `Foo%40Bar%40` instead of `Foo@Bar@`. So if I want to use an `url`, I must use a password with only "standard" characters.

That's a wrong behavior : the user info in an URI can contain special characters that are URL-encoded, and so they must be decoded before using them to connect to the database.

### PR DESCRIPTION
**Pull request by cedx**
Fixes #3716.
Does it require a test?

### CODE REVIEW

File: src/driver/DriverUtils.ts [hunk 1]
```diff
@@ -61,8 +61,8 @@ export class DriverUtils {
         return {
             type: type,
             host: host,
-            username: username,
-            password: password,
+            username: decodeURIComponent(username),
+            password: decodeURIComponent(password),
```
Comment (pleerock) [inline src/driver/DriverUtils.ts]: are you sue other parameters like host shouldn't be decoded as well?

**Comment by pleerock** [comment]
ideally we need a test but not sure if we can setup integration test. So we probably have to make a unit test.

File: src/driver/DriverUtils.ts (hunk 1 above)
Comment (cedx) [reply src/driver/DriverUtils.ts]: In theory, we should decode every part that contains "reserved" characters (cf. https://en.wikipedia.org/wiki/Percent-encoding). In practice, in the 15 years of my career, I haven't see any hostname or database name using reserved characters.

**Comment by cedx** [comment]
I added some unit tests.

**Comment by pleerock** [comment]
Thank you for your contribution!

**Pull request merged**